### PR TITLE
feat(linter): eslint/sort-keys 

### PR DESCRIPTION
**Pull request by Goldziher**
This PR adds the `eslint/sort-keys` implementation as part of #479 

Note: two tests are commented out. I was not able to model this behavior properly. I would be happy to have some assistance there.

### CODE REVIEW

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs [hunk 1]
```diff
@@ -0,0 +1,1077 @@
...
+                    sorted.reverse();
+                }
+
+                sorted_property_groups[i] = sorted;
+            }
+
+            let is_sorted = all(property_groups.iter().zip(&sorted_property_groups), |(a, b)| a == b);
+
+            if !is_sorted {
+                ctx.diagnostic(
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/sort_keys.rs:185]: Use a diagnostic factory function. See other rules for examples.

**Comment by DonIsaac** [comment]
Thanks for your contribution to oxc! This looks pretty good, just a few things to polish off.

File: .gitignore [hunk 2]
```diff
@@ -27,3 +27,5 @@ tasks/coverage/babel/
 tasks/coverage/test262/
 tasks/coverage/typescript/
 tasks/prettier_conformance/prettier/
+
+.idea/
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply .gitignore:31]: Ok, why?

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs [hunk 3]
```diff
@@ -0,0 +1,1077 @@
...
+    /// ```js
+    /// FIXME: Tests will fail if examples are missing or syntactically incorrect.
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// FIXME: Tests will fail if examples are missing or syntactically incorrect.
+    /// ```
+    SortKeys,
+    nursery, // TODO: change category to `correctness`, `suspicious`, `pedantic`, `perf`, `restriction`, or `style`
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:62]: Ok. Which one would this rule be? Style?

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs [hunk 4]
```diff
@@ -0,0 +1,1077 @@
...
+             // See <https://oxc.rs/docs/contribute/linter.html#rule-category> for details
+
+    pending  // TODO: describe fix capabilities. Remove if no fix can be done,
+             // keep at 'pending' if you think one could be added but don't know how.
+             // Options are 'fix', 'fix_dangerous', 'suggestion', and 'conditional_fix_suggestion'
+);
+
+impl Rule for SortKeys {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let config_array = match value.as_array() {
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:72]: 👍🏼

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs (hunk 1 above)
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:185]: 👍🏼

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs [hunk 5]
```diff
@@ -0,0 +1,1077 @@
...
+            Some(v) => v,
+            None => {
+                // we follow the eslint defaults
+                return Self(Box::new(SortKeysOptions {
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:76]: 👍🏼

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:62]: im proceeding for now with sytle. Lemme know if something else is better.

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs [hunk 6]
```diff
@@ -0,0 +1,1077 @@
...
+    ///
+    ///
+    /// ### Why is this bad?
+    ///
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```js
+    /// FIXME: Tests will fail if examples are missing or syntactically incorrect.
```
[the remark this answers is not in the document]
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:54]: 👍🏼

File: .gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (Goldziher) [reply .gitignore:31]: done. 

I simply don't commit it - but I do commit it to other projects, which are not open source, so I am not globally ignoring it. I would recommend its ignored in OXC instead.

**Comment by Goldziher** [comment]
> Thanks for your contribution to oxc! This looks pretty good, just a few things to polish off.

Great. I addressed all comments.

File: Cargo.toml [hunk 7]
```diff
@@ -192,7 +192,7 @@ ignored = ["napi", "oxc_transform_napi", "prettyplease"]
 [profile.dev]
 # Disabling debug info speeds up local and CI builds,
 # and we don't rely on it for debugging that much.
-debug = false
+# debug = false
```
Comment (DonIsaac) [inline Cargo.toml:195]: Please undo this

File: .gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (rzvxa) [reply .gitignore]: Feel free to mark this conversation as resolved, We can add it in a follow-up if Boshen gives a 👍

@DonIsaac

File: Cargo.toml (hunk 7 above)
Comment (Goldziher) [reply Cargo.toml:195]:
```suggestion
```

Comment (Goldziher) [reply Cargo.toml]: unintentional. Reverted. Note- this is an issue for running a debugger in intellij.

Comment (DonIsaac) [reply Cargo.toml]: Yup, same here for me in VSCode. I simply don't commit it; it helps keep CI nice and speedy 🏃

File: .gitignore (hunk 2 above)
[the remark this answers is not in the document]
Comment (DonIsaac) [reply .gitignore]: Boshen has expressed that editor-specific files should not be included in oxc's .gitignore. I don't recall his reasoning, but I know he doesn't want it there.

**Review by DonIsaac — APPROVED** [review]

**Comment by DonIsaac** [comment]
@rzvxa feel free to merge this when CI is passing and you are ready.

**Comment by rzvxa** [comment]
@Goldziher It seems like you've done the hard part, Is there anything that I can help with to push this PR past the finish line?

To fix the CI issues you can run `just r` on your local machine.
For updating snapshots you should do the following:

1. `cargo insta test`
2. `cargo insta review`
3. accept or reject snapshot updates
3. commit the updated snapshots.

The snapshot you've committed is an intermediate one. `Insta` on tests creates `new` snapshots and after review, it would remove it in favor of the "actual" snapshot.

The one with `new` in the name is for local use in the review process and shouldn't be committed.

**Comment by Goldziher** [comment]
> @Goldziher It seems like you've done the hard part, Is there anything that I can help with to push this PR past the finish line?
> 
> To fix the CI issues you can run `just r` on your local machine. For updating snapshots you should do the following:
> 
> 1. `cargo insta test`
> 2. `cargo insta review`
> 3. accept or reject snapshot updates
> 4. commit the updated snapshots.
> 
> The snapshot you've committed is an intermediate one. `Insta` on tests creates `new` snapshots and after review, it would remove it in favor of the "actual" snapshot.
> 
> The one with `new` in the name is for local use in the review process and shouldn't be committed.

thanks, its done.

**Comment by rzvxa** [comment]
There are still failing CI tasks, Have you tried using `cargo lint -- -D warnings` or `just r` commands?

**Comment by Goldziher** [comment]
> There are still some failing CI tasks, Have you tried using `cargo lint -- -D warnings` or `just r` commands?

I'll try running it tomorrow, cheers

**Comment by Goldziher** [comment]
Hi, @rzvxa - i fixed all the clippy issues and ensured the tests are all passing. I am getting one error running `just r` that I don't quite understand:

```text
est linter::rules::test::test_doc_pages ... FAILED
test linter::rules::test::test_rules_table ... ok

failures:

---- linter::rules::test::test_doc_pages stdout ----
  x Expected a semicolon or an implicit semicolon after a statement, but found
  | none
   ,-[Pedantic/eslint/sort-keys/code-snippet:1:10]
 1 | let myObj {
   :          ^
 2 |   c: 1,
   `----
  help: Try insert a semicolon here

thread 'linter::rules::test::test_doc_pages' panicked at tasks/website/src/linter/rules/test.rs:94:64:
called `Result::unwrap()` on an `Err` value: "1 errors occurred while parsing Pedantic/eslint/sort-keys/code-snippet.jsx"
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    linter::rules::test::test_doc_pages

test result: FAILED. 5 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.16s

error: test failed, to rerun pass `-p website --lib`
error: Recipe `test` failed on line 67 with exit code 101
error: Recipe `ready` failed on line 27 with exit code 101
```

I don't understand what semi-colon it wants from me to add - the JS is find and there is a semi-colon at the end of the statement.

**Comment by rzvxa** [comment]
You are missing an equal sign there:

```js
let obj = { /* ... */ }
```
Since there is no equal sign it interprets it as two expressions e.g.
```js
let obj; { /* ... */ }
```
And says to add a semicolon to be more explicit.

**Comment by Goldziher** [comment]
> You are missing an equal sign there:
> 
> ```js
> let obj = { /* ... */ }
> ```
> 
> Since there is no equal sign it interprets it as two expressions e.g.
> 
> ```js
> let obj; { /* ... */ }
> ```
> 
> And says to add a semicolon to be more explicit.

🤦🏼 

All fixed and everything is green locally. Thanks for your help!

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs [hunk 8]
```diff
@@ -0,0 +1,1093 @@
...
+    ///   c: 1,
+    /// };
+    /// ```
+    SortKeys,
+    pedantic, // TODO: change category to `correctness`, `suspicious`, `pedantic`, `perf`, `restriction`, or `style`
+             // See <https://oxc.rs/docs/contribute/linter.html#rule-category> for details
+
+    pending  // TODO: describe fix capabilities. Remove if no fix can be done,
+             // keep at 'pending' if you think one could be added but don't know how.
+             // Options are 'fix', 'fix_dangerous', 'suggestion', and 'conditional_fix_suggestion'
```
Comment (jelly) [inline crates/oxc_linter/src/rules/eslint/sort_keys.rs:89]: The TODO comments can be removed as they placeholder has already been replaced.

Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:89]: fixed

**Comment by Goldziher** [comment]
@rzvxa is this PR done?

**Comment by DonIsaac** [comment]
@rzvxa appears busy, I'll review this in the morning.

**Comment by Goldziher** [comment]
> @rzvxa appears busy, I'll review this in the morning. 

Anything?

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs [hunk 9]
```diff
@@ -0,0 +1,1089 @@
...
+    /// ```
+    SortKeys,
+    pedantic,
+    pending
+);
+
+impl Rule for SortKeys {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let Some(config_array) = value.as_array() else {
+            return Self(Box::default());
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/sort_keys.rs:91]: Consider `Self::default()` for clarity

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs [hunk 10]
```diff
@@ -0,0 +1,1089 @@
...
+        return Ordering::Less;
+    }
+
+    /* empty keys special case */
+    if a.is_empty() && b.starts_with('[') || b.is_empty() && a.starts_with('[') {
+        return Ordering::Equal;
+    }
+
+    let len = a.len().min(b.len());
+    let a_chars: Vec<char> = a.chars().take(len).collect();
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/sort_keys.rs:215]: Can you zip these two iterations together without collecting?

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs [hunk 11]
```diff
@@ -0,0 +1,1089 @@
...
+        }
+
+        if !a_char.is_alphanumeric() && b_char.is_alphanumeric() {
+            return Ordering::Less;
+        }
+
+        return a_char.cmp(b_char);
+    }
+
+    a.cmp(b)
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/sort_keys.rs:251]: Can we not return `Ordering::Equal` here?

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs [hunk 12]
```diff
@@ -0,0 +1,1089 @@
...
+        ("var obj = {[``]:1, a:2}", Some(serde_json::json!([]))), // { "ecmaVersion": 6 },
+        ("var obj = {_:2, a:1, b:3} // default", Some(serde_json::json!([]))),
+        ("var obj = {a:1, b:3, c:2}", Some(serde_json::json!([]))),
+        ("var obj = {a:2, b:3, b_:1}", Some(serde_json::json!([]))),
+        ("var obj = {C:3, b_:1, c:2}", Some(serde_json::json!([]))),
+        ("var obj = {$:1, A:3, _:2, a:4}", Some(serde_json::json!([]))),
+        ("var obj = {1:1, '11':2, 2:4, A:3}", Some(serde_json::json!([]))),
+        ("var obj = {'#':1, 'Z':2, À:3, è:4}", Some(serde_json::json!([]))),
+        ("var obj = { [/(?<zero>0)/]: 1, '/(?<zero>0)/': 2 }", Some(serde_json::json!([]))), // { "ecmaVersion": 2018 },
+        // ("var obj = {a:1, b:3, [a + b]: -1, c:2}", Some(serde_json::json!([]))), // { "ecmaVersion": 6 },
```
Comment (DonIsaac) [inline crates/oxc_linter/src/rules/eslint/sort_keys.rs:341]: What happened here?

**Comment by DonIsaac** [comment]
Apologies for the delay, I got sidetracked 😅

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs (hunk 12 above)
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:341]: see the PR description - there are two cases where I wasn't able to match the test cases.

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs (hunk 11 above)
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:251]: I am not 100% sure, thats why I kept it like this.

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs (hunk 10 above)
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:215]: done

File: crates/oxc_linter/src/rules/eslint/sort_keys.rs (hunk 9 above)
Comment (Goldziher) [reply crates/oxc_linter/src/rules/eslint/sort_keys.rs:91]: done

**Comment by Goldziher** [comment]
@DonIsaac I addressed your comments and made the changes. I cannot resolve the merge conflict - please push a fix to this branch.

**Comment by DonIsaac** [comment]
> @DonIsaac I addressed your comments and made the changes. I cannot resolve the merge conflict - please push a fix to this branch.

The conflicts aren't too bad, I believe in you 😄

**Comment by Goldziher** [comment]
> > @DonIsaac I addressed your comments and made the changes. I cannot resolve the merge conflict - please push a fix to this branch.
> 
> The conflicts aren't too bad, I believe in you 😄 

That's not the issue. Ill need to add a second remote now, but I'll do it.

**Comment by Goldziher** [comment]
Should be good to go @DonIsaac

**Comment by DonIsaac** [comment]
Once CI is passing we can merge this

**Pull request merged**